src/StateService/StatePlugin.RPC.cs

```diff
@@ -0,0 +1,97 @@
+using System.Text;
+using System.IO;
+
+namespace Neo.Plugins.StateService
+{
+    public partial class StatePlugin
+    {
+        [RpcMethod]
+        public JObject GetStateRoot(JArray _params)
+        {
+            uint index = uint.Parse(_params[0].AsString());
+            StateRoot state_root = StateStore.Singleton.StateRoots.TryGet(index);
+            if (state_root is null)
+                throw new RpcException(-100, "Unknown root hash");
+            else
+                return state_root.ToJson();
+        }
+
+        [RpcMethod]
+        public JObject GetProof(JArray _params)
+        {
+            UInt256 root_hash = UInt256.Parse(_params[0].AsString());
+            if (!Settings.Default.FullState)
+            {
+                if (StateStore.Singleton.CurrentLocalRootHash != root_hash) throw new RpcException(-100, "Unknown root hash");
+            }
+            UInt160 script_hash = UInt160.Parse(_params[1].AsString());
+            byte[] key = _params[2].AsString().HexToBytes();
+            var contract = Blockchain.Singleton.View.Contracts.TryGet(script_hash);
+            if (contract is null) throw new RpcException(-100, "Unknown contract");
+            StorageKey skey = new StorageKey
+            {
+                Id = contract.Id,
+                Key = key,
+            };
+            HashSet<byte[]> proof = StateStore.Singleton.GetProof(root_hash, skey);
+            if (proof is null) throw new RpcException(-100, "Unknown value");
+
+            using MemoryStream ms = new MemoryStream();
+            using BinaryWriter writer = new BinaryWriter(ms, Encoding.UTF8);
+
+            writer.WriteVarBytes(skey.ToArray());
+            writer.WriteVarInt(proof.Count);
+            foreach (var item in proof)
+            {
+                writer.WriteVarBytes(item);
+            }
+            writer.Flush();
+
+            return ms.ToArray().ToHexString();
+        }
+
+        [RpcMethod]
+        public JObject VerifyProof(JArray _params)
+        {
+            UInt256 root_hash = UInt256.Parse(_params[0].AsString());
+            byte[] proof_bytes = _params[1].AsString().HexToBytes();
+            var proof = new HashSet<byte[]>();
+
+            using MemoryStream ms = new MemoryStream(proof_bytes, false);
+            using BinaryReader reader = new BinaryReader(ms, Encoding.UTF8);
+
+            var key = reader.ReadVarBytes(MPTNode.MaxKeyLength);
+            var count = reader.ReadVarInt();
+            for (ulong i = 0; i < count; i++)
+            {
+                proof.Add(reader.ReadVarBytes());
+            }
+
+            var skey = key.AsSerializable<StorageKey>();
+            var sitem = MPTTrie<StorageKey, StorageItem>.VerifyProof(root_hash, skey, proof);
+            if (sitem is null) throw new RpcException(-100, "Verification failed");
+            return sitem.Value.ToHexString();
+        }
+
+        [RpcMethod]
+        public JObject VoteStateRoot(JArray _params)
```

Comment:
I see now, the idea is to move from tasks in `ConsensusService` to tasks here on the MPTService done by `DesignedNodes`, right?

Reply:
Yes.